### DISCUSSION TITLE
raw request in app router route.ts file

### DISCUSSION THREAD

**Discussion by artooras** [vercel/next.js#62941] [App Router]
I'm using NextJS for building a Shopify App. Before migrating to the app router, in my pages/api route I had the following code which was **working**:

```
export default async function handler(req: NextApiRequest) {
    
  ...
  
  const sessionId = await shopify.session.getCurrentId({
    isOnline: false,
    rawRequest: req
  })

  ...
```

Once I migrated to the the app router, I now have a route.ts file in which I execute similar code:

```
export async function POST(req: Request) {

  ...

  const sessionId = await shopify.session.getCurrentId({
    isOnline: false,
    rawRequest: req
  })

  ...
```

but this time, instead of a valid `sessionId`, I get an error saying that `authorization header is missing` in the request. Now, the header is actually there, which I am able to access through:

```
console.log(req.headers.get('Authorization'));
```

However, as the `shopify.session.getCurrentId` suggests, it expects a `rawRequest`, while in the `route.ts` the instance of the `req` is that of NextRequest. And somehow the shopify function is not able to access the headers properly.

Is there a way to have the `req` as a simple/raw request?

Thanks!

**Comment by icyJoseph** [comment]
Hi,

The issue here is that Pages Router used a different request API. That one basically extended IncomingRequest. 

And in that API, headers is a so-called plain JavaScript object.

- https://nodejs.org/docs/latest-v16.x/api/http.html#messageheaders

If we look at the API you are using:

- https://github.com/Shopify/shopify-api-js/blob/main/packages/shopify-api/lib/session/session-utils.ts#L29

We see that it does try to index/look-up into a plain JS Object.

However, in App Router, Route Handlers implement the WebAPI Request object. And in that API, headers extend the Map API:

- https://developer.mozilla.org/en-US/docs/Web/API/Request/headers

Which means that to access headers, you have to use the `get(key)` method.

I'd say that shopify needs to widen the header types they look after. Something like `headers instance of Headers` could help their code check what kind of API they are working with.

As for your case, maybe you need to copy the Request object into a `new http.IncomingMessage` instance? Like

```js
// pseudocode
// req is the App Router Request input
let headers = req.headers // copy reference to headers from App Route request
let incomingReq = new http.IncomingMessage(req);
incomingReq.headers = Object.fromEntries(req.entries())
```

And then pass that to the `rawRequest`. Alternatively, maybe the Shopify API has some way to handle Request, rather than IncomingMessage?

> I don't expect the pseudocode above to work, and it probably has numerous pitfalls hidden

**Reply by artooras** [reply]
Hi @icyJoseph, thanks a lot for such a detailed response! Where should the `http` object be imported from to be able to create a `new http.IncomingRequest(req)`?

**Reply by icyJoseph** [reply]
That comes from `import http from 'http'`, that's a Node.js module.

**Reply by artooras** [reply]
VSCode strangely has not suggested this... :) However, it now states that there is no `IncomingRequest` property on `http`. If I try to ignore it and still run the code, I get an error stating that `http.IncomingRequest` is not a constructor.

**Reply by icyJoseph** [reply]
Oh that's because you want `http.IncomingMessage` I wrote it wrong in the pseudocode. :) Corrected it now.

**Reply by artooras** [reply]
Yes, that's better! But now it does not accept `req` as an argument... :) Says it has to be of type `Socket`.

**Reply by icyJoseph** [reply]
Yeah, like I said, you are probably looking at a couple of hours work to copy stuff from req into IncomingMessage ~ only to get shopify to believe that it got a rawRequest...

I'd say, this is more of an issue or lack of support, within that API.

**Reply by artooras** [reply]
I think I'll have to revert to using the pages router for api routes in the meantime... Thanks a lot for your help trying to figure this out!
